### DISCUSSION TITLE
Unexpected "MBEDTLS_ERR_MPI_ALLOC_FAILED" error while connecting via MQTTs with frozen firmware

### DISCUSSION THREAD

**Discussion by djs0109** [micropython/micropython#9947] [Using MicroPython]
Hi everyone,

I am developing a project for ESP32, which needs to connect to MQTT Broker via TLS/SSL. As usual, I first develop the project with the file system, e.g. I upload my codes to the board and let them run on RAM. Now I want to build a custom firmware, i.e. to freeze my project to a `.bin` file. I have followed the official tutorials and managed to do this. But I got some weird errors.

**Problem**
I got the error `OSError: (-17040, 'MBEDTLS_ERR_RSA_PUBLIC_FAILED+MBEDTLS_ERR_MPI_ALLOC_FAILED')` when my ESP32 trying to establish a MQTTs connection. However, this error only shows up when I use my custom firmware. If I use the official firmware and upload all my code to the board, it can successfully connect to the Broker.

**Debug**
For debugging, I have checked the free memory (with `gc.mem_free()` ) in both cases before establishing a MQTTs connection. 
1. With official firmware and uploaded codes: `free_memory = 6352`
2. With custom firmware and frozen codes: `free_memory = 71072`

This result makes me even more confusing, because it shows that there are more free memory when I use custom firmware. However the error I got indicates that the memory is not enough ...

**Comment by djs0109** [comment]
I may find the cause of my problem. I have downloaded the latest official firmware `V1.19.1` from [here](https://micropython.org/download/esp32/). However, my custom firmware is built from the `master` branch of micropython, because I can not include the `micropython-lib` under tag `1.19.1`. Maybe someone can show me the right way to build a complete `V1.19.1` firmware, and I will test again with that.

**Reply by jimmo** [reply]
Sorry I'm not quite sure what you mean here.

But, I think your problem is likely related to https://github.com/micropython/micropython/issues/8940

If the 1.19.1 firmware is working, but your self-compiled firmware is not, then the likely reason is that the official releases use IDF v4.2 to work around the above issue, whereas you might be using v4.4? Can you try using IDF v4.2?

**Reply by djs0109** [reply]
Thank you jimmo, I think I have used IDF v4.0.2 as [here](https://github.com/micropython/micropython/tree/master/ports/esp32). I will definitely try with v4.2

**Reply by djs0109** [reply]
@jimmo that solve my problem!

**Comment by djs0109** [comment]
I meet a another weird problem right now. The `OSError: (-17040, 'MBEDTLS_ERR_RSA_PUBLIC_FAILED+MBEDTLS_ERR_MPI_ALLOC_FAILED')` sometimes still occur while connecting to MQTT Broker via TLS for the FIRST time. Once the first connection is established, my program works relatively stable (one device has been running for 2 weeks without problem).

My question is, isn't the available memory space relatively constant after every restart? How can this problem just come up with a certain probability? As extra information, I also do a `gc.collect()` before calling `mqtt.connect()`

**Reply by abuvanth** [reply]
same problem 

```
import urequests as rt

r = rt.get("https://cyberfly.io")
print(r.text)```

Traceback (most recent call last):
  File "<stdin>", line 11, in <module>
  File "urequests.py", line 180, in get
  File "urequests.py", line 93, in request
OSError: (-17040, 'MBEDTLS_ERR_RSA_PUBLIC_FAILED+MBEDTLS_ERR_MPI_ALLOC_FAILED')
